source/common/http/conn_manager_impl.cc

```diff
@@ -484,6 +504,25 @@ void ConnectionManagerImpl::doConnectionClose(
   }
 }
 
+void ConnectionManagerImpl::createStartDrainTimer(std::chrono::milliseconds drain_delay) {
+  start_drain_timer_ = read_callbacks_->connection().dispatcher().createTimer([this]() -> void {
+    if (drain_state_ != DrainState::NotDraining) {
+      return;
+    }
+    if (!codec_) {
```

Comment:
Should createStartDrainTimer ever be called when codec is no longer present?  Should we be creating timers for drain after doing the doConnectionClose a few lines down?

I think there's some inconsistencies in implementation that add to complexity.  Would be good to avoid a call to createStartDrainTimer once the codec has been deleted, and better understand cases where this situation currently happens.

Reply:
I've moved the `doConnectionClose` outside of the timer creation. My original implementation hadn't fully considered the lifespan of the `codec_`. Reading through the code I believe you're right that there isn't any reason to wrap the logic in a timer if the `codec_` is null when creating the timer.